core/commands/storage.go

```diff
@@ -438,12 +448,78 @@ func getValidHost(retryQueue *storage.RetryQueue) (*storage.HostNode, error) {
 				return nil, err
 			}
 		} else {
+			// find peer
+			pi, err := findPeer(ctx, n, nextHost.Identity)
+			if err != nil {
+				// cannot find peer, remove from queue permanently
+				log.Error(err)
+				continue
+			}
+			if err := api.Swarm().Connect(ctx, *pi); err != nil {
+				log.Error("fail swarm connect first time")
+				// force connect again
+				if err := api.Swarm().Connect(ctx, *pi); err != nil {
+					log.Error("fail swarm connect second time")
+					// in test mode, change the ip to 0.0.0.0
+					if test {
+						if err := changeAddress(pi); err != nil {
+							log.Error(err)
```

Comment:
on err, you should skip `Connect` and go to the IncrementFail directly. Try to organize the fail path a little more.

for example (rough idea)

```
var err error
if err = api.Swarm.Connect(...); err != nil {
    <change address>
    if err == nil {
        err = api.Swarm.Connect(...)
    }
}
if err != nil {
   <increment fail>
   <offer>
   if err != nil { return ... }
   continue
}
```

Reply:
So you mean give local connect also 2 times connect?